### PR TITLE
Fix broken merge in CUP compile event. Fixes #3773.

### LINKED ISSUE #3773: Support for Cognito User Pool Triggers isn't working.
# This is a Bug Report
## Description
Support for Cognito User Pool Triggers isn't working.

For bug reports:
* What went wrong?
Cannot link lamda to cognito user pool triggers
* What did you expect should have happened?
Cognito user pool triggers should have been updated with the correct lambda.
* What was the config you used?
```yaml
service: hc-cognito-service # NOTE: update this with your service name

provider:
  name: aws
  runtime: nodejs6.10

functions:
  customMessageHandler:
    handler: custom-message-handler.handle
    events:
      - cognitoUserPool:
          pool: <my pool Id>
          trigger: CustomMessage
```
* What stacktrace or error message from your provider did you see?
No error messages were seen.

For feature proposals:
* What is the use case that should be solved. The more detail you describe this in the easier it is to understand for us.
NA
* If there is additional config how would it look
NA

Similar or dependent issues:

## Additional Data

* ***Serverless Framework Version you're using***: 1.15.2
* ***Operating System***: MacOS 10.12.5
* ***Stack Trace***: None
* ***Provider Error messages***: None

### PR DESCRIPTION
**Pull request by hassankhan**
## What did you implement:

Closes #3773

## How did you implement it:

Add a check to see if an override exists in the `Resources` section of `serverless.yml`. If one does exist, then merge the overrides into the base template.

## How can we verify it:

By running the unit tests. I'll try and get some time to properly test it, but from what I can tell it was mainly a CF template generation issue anyway.

## Todos:

- [x] Write tests
- [x] ~~Write documentation~~
- [x] Fix linting errors
- [x] Make sure code coverage hasn't dropped
- [x] Provide verification config / commands / resources
- [x] Enable "Allow edits from maintainers" for this PR
- [x] Update the messages below

***Is this ready for review?:*** YES
***Is it a breaking change?:*** NO

### CODE REVIEW

**Comment by hassankhan** [comment]
Thanks for the review @pmuens! What would be the best way to achieve this then? Should I be looking at `mergeCustomProviderResources` instead?

File: lib/plugins/aws/package/compile/events/cognitoUserPool/index.test.js [hunk 1]
```diff
@@ -264,6 +264,53 @@ describe('AwsCompileCognitoUserPoolEvents', () => {
...
+      };
+
+
+      awsCompileCognitoUserPoolEvents.compileCognitoUserPoolEvents();
+
+      expect(awsCompileCognitoUserPoolEvents.serverless.service.provider
+        .compiledCloudFormationTemplate.Resources
+        .CognitoUserPoolMyUserPool.Type
+      ).to.equal('AWS::Cognito::UserPool');
+      expect(Object.keys(awsCompileCognitoUserPoolEvents.serverless.service.provider
```
Comment (HyperBrain) [inline lib/plugins/aws/package/compile/events/cognitoUserPool/index.test.js]: The length checks can be simplified - to sound more natural:
```
expect(_.keys(awsCompileCognitoUserPoolEvents.serverless.service.provider
        .compiledCloudFormationTemplate.Resources
        .CognitoUserPoolMyUserPool.Properties))
    .to.have.lengthOf(6);
```

Additionally, lodash should be preferred over the native functions, to ensure compatibility over multiple Node versions that could be used by the user (e.g. `_.keys()`, `_.assign()`, etc.)

File: lib/plugins/aws/package/compile/events/cognitoUserPool/index.test.js [hunk 2]
```diff
@@ -264,6 +264,53 @@ describe('AwsCompileCognitoUserPoolEvents', () => {
...
+        .compiledCloudFormationTemplate.Resources
+        .CognitoUserPoolMyUserPool.Properties).length
+      ).to.equal(6);
+      expect(Object.keys(awsCompileCognitoUserPoolEvents.serverless.service.provider
```
Comment (HyperBrain) [inline lib/plugins/aws/package/compile/events/cognitoUserPool/index.test.js]: Same here.

File: lib/plugins/aws/package/compile/events/cognitoUserPool/index.js [hunk 3]
```diff
@@ -115,6 +116,12 @@ class AwsCompileCognitoUserPoolEvents {
         DependsOn,
       };
 
+      // If overrides exist in `Resources`, merge them in
+      if (_.has(this.serverless.service.resources, userPoolLogicalId)) {
```
Comment (HyperBrain) [inline lib/plugins/aws/package/compile/events/cognitoUserPool/index.js]: The merge of `DependsOn` values present in the overrides is not handled correctly. If the override specifies a `DependsOn: "ResA"` or `DependsOn: [ "ResB", "ResC" ]`, the resulting DependsOn shoud be a concatenation of the of the values in the override and of the resource itself. As far as I know, `_.merge()` does not merge arrays but performs an assign.
You have to merge the dependencies in the override manually here. And, important, DependsOn can either be a string, or an array. The implementation should handle both cases.

**Comment by pmuens** [comment]
> Should I be looking at mergeCustomProviderResources instead?

Yes, I think the solution is to hook into the `mergeCustomProviderResources` lifecycle event and run the functionality when this lifecycle event kicks in.

File: lib/plugins/aws/package/compile/events/cognitoUserPool/index.test.js (hunk 1 above)
Comment (hassankhan) [inline lib/plugins/aws/package/compile/events/cognitoUserPool/index.test.js]: I'll replace those in all the test cases in this file 👍

**Comment by hassankhan** [comment]
I think I've done what you requested, @pmuens. Just a little curious, shouldn't other events that _could_ have custom resources defined specify the same hook? For example, I used the S3 event as inspiration, and that doesn't seem to have a `mergeCustomProviderResources` hook.

**Comment by pmuens** [comment]
> I think I've done what you requested, @pmuens. Just a little curious, shouldn't other events that could have custom resources defined specify the same hook? For example, I used the S3 event as inspiration, and that doesn't seem to have a mergeCustomProviderResources hook.

Thanks for the updates @hassankhan 👍 I'll review them ASAP.

The reason has smth. todo with the legacy code base. The `s3` event was the first event integration we ever wrote. There was no idea / need to overwrite / merge other S3 bucket definitions with the one compiled by the plugin.

File: lib/plugins/aws/package/compile/events/cognitoUserPool/index.js [hunk 4]
```diff
@@ -20,19 +20,21 @@ class AwsCompileCognitoUserPoolEvents {
 
     this.hooks = {
       'package:compileEvents': this.compileCognitoUserPoolEvents.bind(this),
+      'aws:package:finalize:mergeCustomProviderResources': this
```
Comment (pmuens) [inline lib/plugins/aws/package/compile/events/cognitoUserPool/index.js]: We should update this to use the `after:package:finalize` lifecycle.

The main reason for that is that the lifecycle event which is used here does not ensure order (because we want to merge our Cognito User Pool resources `after` the `merge` has happened) and the `aws:x` lifecycle events are not "official" Serverless lifecycle events.

The Serverless lifecycle events which are not inner lifecycle events are officially marked as non-breaking since they're less likely to be changed.

So in conclusion this should be updated to: `after:package:finalize`

File: lib/plugins/aws/package/compile/events/cognitoUserPool/index.js [hunk 5]
```diff
@@ -80,51 +82,59 @@ class AwsCompileCognitoUserPoolEvents {
...
-      };
+      },
+    };
+  }
 
-      const userPoolCFResource = {
-        [userPoolLogicalId]: userPoolTemplate,
-      };
+  compileCognitoUserPoolEvents() {
+    const { cognitoUserPoolTriggerFunctions, userPools } = this.findUserPoolsAndFunctions();
```
Comment (pmuens) [inline lib/plugins/aws/package/compile/events/cognitoUserPool/index.js]: Deconstruction will break in Node.js 4 which we still support.

File: lib/plugins/aws/package/compile/events/cognitoUserPool/index.js [hunk 6]
```diff
@@ -159,6 +169,41 @@ class AwsCompileCognitoUserPoolEvents {
         permissionCFResource);
     });
   }
+
+  mergeWithCustomResources() {
+    const { cognitoUserPoolTriggerFunctions, userPools } = this.findUserPoolsAndFunctions();
```
Comment (pmuens) [inline lib/plugins/aws/package/compile/events/cognitoUserPool/index.js]: Deconstruction will break in Node.js 4 which we still support.

**Comment by hassankhan** [comment]
Sorry for the delayed response, been super busy with work commitments. I've updated as requested, totally forgot Node 4 is still supported 😁

**Comment by pmuens** [comment]
> Sorry for the delayed response, been super busy with work commitments. I've updated as requested, totally forgot Node 4 is still supported 😁

Thanks for getting back and updating @hassankhan 💯 

We'll review this again in the upcoming week!

No hurry / pressure. We got you covered and help you if you've got some time restrictions 💪 (thanks for taking time to work on this!)

Yes, Node 4 is still around and some framework users are still using it. I'd also like to use all the nifty new features in our codebase, but we need to wait a little bit more until we get there 😄

**Comment by hassankhan** [comment]
Apologies for the late reply, this PR actually fixes overriding a resource created when defining a trigger event.

Using a `serverless.yml` as follows will create a User Pool and attach the Lambda to it (this works):

```yml
functions:
  SignupTrigger:
    handler: SignupTrigger.default
    events:
      - cognitoUserPool:
        pool: MyUserPool
        trigger: PreSignUp
```

The problem is when you want to override properties of the User Pool resource, which you almost definitely will want to do:

```yml
functions:
  SignupTrigger:
    handler: SignupTrigger.default
    events:
      - cognitoUserPool:
        pool: MyUserPool
        trigger: PreSignUp

resources:
  CognitoUserPoolMyUserPool:
    Type: "AWS::Cognito::UserPool"
    Properties:
      ...
```

**Comment by pmuens** [comment]
> Apologies for the late reply, this PR actually fixes overriding a resource created when defining a trigger event.

No worries @hassankhan 👍

Thanks you very much for the in-depth explanation with example service files to verify the PR 👍 and thanks again for your work on the PR 💯 .

**Comment by Nosajool** [comment]
I am having trouble creating a trigger for a pool defined in `serverless.yml`. Would this PR fix this?

```yaml
Resources:
    UserPool:
      Type: "AWS::Cognito::UserPool"
      Properties:
...

functions:
  autoConfirmUser:
    handler: src/user/confirm.main
    events:
      - cognitoUserPool:
        pool:
          Ref: UserPool
        trigger: PreSignUp
...
```

**Comment by deathweaselx86** [comment]
Just started reviewing :) <br><br>:octocat: *Sent from [GH](http://nodegh.io).*

**Comment by hassankhan** [comment]
Hi, what's the status on this? /cc @pmuens @HyperBrain

**Comment by HyperBrain** [comment]
@hassankhan Sorry, that this one got "lost" somewhere in the pile of PRs 😃 . I'll check it again in the next days.

**Comment by hassankhan** [comment]
Much appreciated, @HyperBrain 😄

**Comment by horike37** [comment]
@HyperBrain @hassankhan 
Seems like there's been no progress on this PR.
Will you continue working on this? ,or closing once?

**Comment by hassankhan** [comment]
It's ready to be merged, as far as I remember.

**Comment by hassankhan** [comment]
🍾

**Comment by HyperBrain** [comment]
@hassankhan I just missed the PR with the open change request - sorry for that 😄 .

**Comment by supun19** [comment]
`PostConfirmation:
    handler: signup/create.create
    events:
      - cognitoUserPool:
          pool: 
            Ref: UserPool
          trigger: PostConfirmation`


`UserPool:
      Type: "AWS::Cognito::UserPool"
      Properties:
        UserPoolName: consult_user_pool_dev
        AutoVerifiedAttributes:
          - email
        MfaConfiguration: OFF
        Schema:
          - AttributeDataType: String
            Name: email
            Required: true

          - AttributeDataType: String
            Name: tenant
            Mutable: true
          - AttributeDataType: String
            Name: mainRole
            Mutable: true
          - AttributeDataType: String
            Name: subRole
            Mutable: true
          - AttributeDataType: String
            Name: phone_numbe
            Mutable: true`

this configuration not set lambda tigger for cognito user pool. where did i do the mistake

**Comment by harleyguru** [comment]
Hey @hassankhan, when this will be fixed permanently?
I am waiting for the fixing of this issue.

**Comment by break-paradigm** [comment]
I'm following the basic Serverless documentation and keep getting the error: 

"The CloudFormation template is invalid: Template format error: Unresolved resource dependencies [CognitoUserPool] in the Resources block of the template" 

Has anyone resolved this issue?
